File: UnitTests/AuthenticationServiceTests.vb
Imports System.Threading.Tasks
Imports Xunit
Imports Moq
Imports Core.Models
Imports Core.Services
Imports Core.Repositories
Imports Microsoft.AspNetCore.Identity

Public Class AuthenticationServiceTests
    Private _userRepositoryMock As Mock(Of IUserRepository)
    Private _passwordHasherMock As Mock(Of IPasswordHasher(Of User))
    Private _authenticationService As AuthenticationService

    Public Sub New()
        _userRepositoryMock = New Mock(Of IUserRepository)()
        _passwordHasherMock = New Mock(Of IPasswordHasher(Of User))()
        _authenticationService = New AuthenticationService(_userRepositoryMock.Object, _passwordHasherMock.Object, "TestJWTKey")
    End Sub

    <Fact>
    Public Async Function RegisterAsync_UserDoesNotExist_CreatesUser() As Task
        ' Arrange
        Dim newUser As New User() With {.Email = "test@test.com"}
        _userRepositoryMock.Setup(Function(x) x.GetUserByEmailAsync(It.IsAny(Of String))).Returns(Task.FromResult(Of User)(Nothing))

        ' Act
        Dim result = Await _authenticationService.RegisterAsync(newUser, "testPassword")

        ' Assert
        _userRepositoryMock.Verify(Function(x) x.AddUserAsync(It.IsAny(Of User)), Times.Once)
        Assert.True(result.IsAuthenticated)
    End Function

    <Fact>
    Public Async Function AuthenticateAsync_CorrectCredentials_ReturnsAuthenticatedUser() As Task
        ' Arrange
        Dim testUser As New User() With {.Email = "test@test.com", .PasswordHash = "hashedPassword"}
        _userRepositoryMock.Setup(Function(x) x.GetUserByEmailAsync(It.IsAny(Of String))).Returns(Task.FromResult(testUser))
        _passwordHasherMock.Setup(Function(x) x.VerifyHashedPassword(It.IsAny(Of User), It.IsAny(Of String), It.IsAny(Of String))).Returns(PasswordVerificationResult.Success)

        ' Act
        Dim result = Await _authenticationService.AuthenticateAsync(New AuthenticateRequest() With {.Email = "test@test.com", .Password = "testPassword"})

        ' Assert
        Assert.True(result.IsAuthenticated)
    End Function
End Class

File: UnitTests/DataAccess/AuthenticationServiceTests.cs
using Xunit;
using Moq;
using Core.Models;
using Core.Repositories;
using Microsoft.AspNetCore.Identity;
using DataAccess.Services;
using System.Threading.Tasks;

namespace UnitTests
{
    public class AuthenticationServiceTests
    {
        private readonly AuthenticationService _authService;
        private readonly Mock<IUserRepository> _mockUserRepo;
        private readonly Mock<IPasswordHasher<User>> _mockPasswordHasher;

        public AuthenticationServiceTests()
        {
            _mockUserRepo = new Mock<IUserRepository>();
            _mockPasswordHasher = new Mock<IPasswordHasher<User>>();
            _authService = new AuthenticationService(_mockUserRepo.Object, _mockPasswordHasher.Object, "test_jwt_key");
        }

        [Fact]
        public async Task RegisterAsync_ShouldReturnAuthenticated_WhenUserIsAddedSuccessfully()
        {
            // Arrange
            var user = new User { Email = "test@test.com" };
            var password = "testPassword";
            _mockUserRepo.Setup(repo => repo.CreateAsync(user)).Returns(Task.CompletedTask);
            _mockPasswordHasher.Setup(hasher => hasher.HashPassword(user, password)).Returns("hashedPassword");

            // Act
            var result = await _authService.RegisterAsync(user, password);

            // Assert
            Assert.True(result.IsAuthenticated);
            Assert.NotNull(result.Token);
            _mockUserRepo.Verify(repo => repo.CreateAsync(user), Times.Once);
            _mockPasswordHasher.Verify(hasher => hasher.HashPassword(user, password), Times.Once);
        }

        [Fact]
        public async Task LoginAsync_ShouldReturnNotAuthenticated_WhenUserDoesNotExist()
        {
            // Arrange
            var email = "nonexistent@test.com";
            var password = "testPassword";
            _mockUserRepo.Setup(repo => repo.GetUserByEmailAsync(It.IsAny<string>())).Returns(Task.FromResult<User>(null));

            // Act
            var result = await _authService.LoginAsync(email, password);

            // Assert
            Assert.False(result.IsAuthenticated);
            _mockUserRepo.Verify(repo => repo.GetUserByEmailAsync(It.IsAny<string>()), Times.Once);
        }

        [Fact]
        public async Task LoginAsync_ShouldReturnAuthenticated_WhenPasswordIsCorrect()
        {
            // Arrange
            var user = new User { Email = "test@test.com", PasswordHash = "hashedPassword" };
            var password = "testPassword";
            _mockUserRepo.Setup(repo => repo.GetUserByEmailAsync(user.Email)).Returns(Task.FromResult(user));
            _mockPasswordHasher.Setup(hasher => hasher.VerifyHashedPassword(user, user.PasswordHash, password)).Returns(PasswordVerificationResult.Success);

            // Act
            var result = await _authService.LoginAsync(user.Email, password);

            // Assert
            Assert.True(result.IsAuthenticated);
            Assert.NotNull(result.Token);
            _mockUserRepo.Verify(repo => repo.GetUserByEmailAsync(user.Email), Times.Once);
            _mockPasswordHasher.Verify(hasher => hasher.VerifyHashedPassword(user, user.PasswordHash, password), Times.Once);
        }

        [Fact]
            public async Task AuthenticateAsync_ShouldReturnAuthenticated_WhenUserCredentialsAreCorrect()
        {
            // Arrange
            var user = new User { Email = "test@test.com", PasswordHash = "hashedPassword" };
            var authRequest = new AuthenticateRequest(user.Email, "testPassword");
            _mockUserRepo.Setup(repo => repo.GetUserByEmailAsync(user.Email)).Returns(Task.FromResult(user));
            _mockPasswordHasher.Setup(hasher => hasher.VerifyHashedPassword(user, user.PasswordHash, "testPassword")).Returns(PasswordVerificationResult.Success);

            // Act
            var result = await _authService.AuthenticateAsync(authRequest);

            // Assert
            Assert.True(result.IsAuthenticated);
            Assert.NotNull(result.Token);
            _mockUserRepo.Verify(repo => repo.GetUserByEmailAsync(user.Email), Times.Once);
        }
    }
}
